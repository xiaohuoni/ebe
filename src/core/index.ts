import { createProjectBuilder } from './generator/ProjectBuilder';
import { createModuleBuilder } from './generator/ModuleBuilder';
import { createDiskPublisher } from './publisher/disk';
import { createZipPublisher } from './publisher/zip';

// 引入说明
import {
  COMMON_CHUNK_NAME,
  CLASS_DEFINE_CHUNK_NAME,
  DEFAULT_LINK_AFTER,
} from './const/generator';

// 引入通用插件组
import esmodule from './plugins/common/esmodule';
import requireUtils from './plugins/common/requireUtils';

import prettier from './postprocessor/prettier';

// 引入全局常用工具
import * as globalUtils from './utils';

import * as CONSTANTS from './const';

export default {
  createProjectBuilder,
  createModuleBuilder,
  solutions: {},
  solutionParts: {},
  publishers: {
    disk: createDiskPublisher,
    zip: createZipPublisher,
  },
  plugins: {
    common: {
      /**
       * 处理 ES Module
       * @deprecated please use esModule
       */
      esmodule,
      esModule: esmodule,
      requireUtils,
    },
  },
  postprocessor: {
    prettier,
  },
  utils: globalUtils,
  chunkNames: {
    COMMON_CHUNK_NAME,
    CLASS_DEFINE_CHUNK_NAME,
  },
  defaultLinkAfter: {
    COMMON_DEFAULT_LINK_AFTER: DEFAULT_LINK_AFTER,
  },
  constants: CONSTANTS,
};

// 一些类型定义
export * from './types';

// 一些常量定义
export * from './const';

// 一些工具函数
export * from './analyzer/componentAnalyzer';
export * from './generator/ChunkBuilder';
export * from './generator/CodeBuilder';
export * from './generator/ModuleBuilder';
export * from './generator/ProjectBuilder';
