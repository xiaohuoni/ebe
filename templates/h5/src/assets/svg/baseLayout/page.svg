<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<g clip-path="url(#clip0_2214_125818)">
<path d="M14.1184 4.01071L10.2756 0.167857C10.1684 0.0607143 10.0238 0 9.87199 0H2.28627C1.9702 0 1.71484 0.255357 1.71484 0.571429V15.4286C1.71484 15.7446 1.9702 16 2.28627 16H13.7148C14.0309 16 14.2863 15.7446 14.2863 15.4286V4.41607C14.2863 4.26429 14.2256 4.11786 14.1184 4.01071ZM12.9684 4.67857H9.6077V1.31786L12.9684 4.67857ZM13.0006 14.7143H3.00056V1.28571H8.39342V5.14286C8.39342 5.34177 8.47243 5.53254 8.61309 5.67319C8.75374 5.81384 8.9445 5.89286 9.14342 5.89286H13.0006V14.7143Z" fill="#1C242E" fill-opacity="0.75"/>
</g>
<defs>
<clipPath id="clip0_2214_125818">
<rect width="16" height="16" fill="white"/>
</clipPath>
</defs>
</svg>
