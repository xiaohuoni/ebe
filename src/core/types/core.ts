import { IProjectSchema } from './page';
import { ResultDir, ResultFile } from './files';
import type { ProjectBuilderInitOptions } from '../generator/ProjectBuilder';
import { IScopeBindings } from '../utils/ScopeBindings';
import { IParseResult } from './intermediate';
import { IContainerInfo } from './';

export enum FileType {
  CSS = 'css',
  SCSS = 'scss',
  LESS = 'less',
  HTML = 'html',
  JS = 'js',
  MJS = 'mjs',
  JSX = 'jsx',
  TS = 'ts',
  MTS = 'mts',
  TSX = 'tsx',
  JSON = 'json',
  MD = 'md',
}

export enum ChunkType {
  AST = 'ast',
  STRING = 'string',
  JSON = 'json',
}

export enum PluginType {
  COMPONENT = 'component',
  UTILS = 'utils',
  I18N = 'i18n',
}

export type ChunkContent = string | any;
export type CodeGeneratorFunction<T> = (content: T) => string;

export interface ICodeChunk {
  type: ChunkType;
  fileType: string;
  name: string;
  subModule?: string;
  content: ChunkContent;
  linkAfter: string[];
  ext?: Record<string, unknown>;
}

export interface IBaseCodeStruct {
  chunks: ICodeChunk[];
  depNames: string[];
}

export interface ICodeStruct extends IBaseCodeStruct {
  ir: any;
  contextData: IContextData;
}

/** 上下文数据，用来在插件之间共享一些数据 */
export interface IContextData extends IProjectBuilderOptions {
  /**
   * 其他自定义数据
   * （三方自定义插件也可以在此放一些数据，建议起个长一点的名称，用自己的插件名做前缀，以防冲突）
   */
  [key: string]: any;

  /**
   * 是否使用了 Ref 的 API (this.$/this.$$)
   * */
  useRefApi?: boolean;
}

export type BuilderComponentPlugin = (
  initStruct: ICodeStruct,
) => Promise<ICodeStruct>;

export type BuilderComponentPluginFactory<T> = (
  config?: T,
) => BuilderComponentPlugin;

export interface IChunkBuilder {
  run: (
    ir: any,
    initialStructure?: ICodeStruct,
  ) => Promise<{ chunks: ICodeChunk[][] }>;
  getPlugins: () => BuilderComponentPlugin[];
  addPlugin: (plugin: BuilderComponentPlugin) => void;
}

export interface ICodeBuilder {
  link: (chunkDefinitions: ICodeChunk[]) => string;
  generateByType: (type: string, content: unknown) => string;
}

export interface ICompiledModule {
  files: ResultFile[];
}

export interface IModuleBuilder {
  generateModule: (input: unknown) => Promise<ICompiledModule>;
  generateModuleCode: (schema: IProjectSchema | string) => Promise<ResultDir>;
  linkCodeChunks: (
    chunks: Record<string, ICodeChunk[]>,
    fileName: string,
  ) => ResultFile[];
  addPlugin: (plugin: BuilderComponentPlugin) => void;
}

/**
 * 引擎对外接口
 *
 * @export
 * @interface ICodeGenerator
 */
export interface ICodeGenerator {
  /**
   * 出码接口，把 Schema 转换成代码文件系统描述
   *
   * @param {(IProjectSchema)} schema 传入的 Schema
   * @returns {ResultDir}
   * @memberof ICodeGenerator
   */
  toCode: (schema: IProjectSchema) => Promise<ResultDir>;
}

export interface ISchemaParser {
  validate: (schema: IProjectSchema) => boolean;
  parse: (
    schema: IProjectSchema | string,
    options?: LXProjectOptions,
  ) => IParseResult;
}

export interface IProjectTemplate {
  slots: Record<string, IProjectSlot>;
  generateTemplate: (
    data: IParseResult,
    postProcessors: PostProcessor[],
  ) => ResultDir | Promise<ResultDir>;
}

export interface IProjectSlot {
  path: string[];
  fileName?: string;
}

export interface IProjectPlugins {
  [slotName: string]: BuilderComponentPlugin[];
}

export interface LXProjectOptions {
  platform: string;
  appId: string;
  pageIdMapping: any;
  busiCompMapping?: any;
  baseUrl?: string;
}
export interface IProjectBuilderOptions {
  /** 是否处于严格模式 (默认：否) */
  inStrictMode?: boolean;

  /**
   * 是否要容忍对 JSExpression 求值时的异常
   * 默认：true
   * 注：如果容忍异常，则会在求值时包裹 try-catch 块，
   *     catch 到异常时默认会抛出一个 CustomEvent 事件里面包含异常信息和求值的表达式
   */
  tolerateEvalErrors?: boolean;

  /**
   * 容忍异常的时候的的错误处理语句块
   * 默认：无
   * 您可以设置为一个语句块，比如：
   *  window.dispatchEvent(new CustomEvent('lowcode-eval-error', { error, expr }))
   *
   * 一般可以结合埋点监控模块用来监控求值异常
   *
   * 其中：
   * - error: 异常信息
   * - expr: 求值的表达式
   */
  evalErrorsHandler?: string;

  /**
   * Hook which is used to customize original options, we can reorder/add/remove plugins/processors
   * of the existing solution.
   */
  customizeBuilderOptions?(
    originalOptions: ProjectBuilderInitOptions,
  ): ProjectBuilderInitOptions;

  // TODO: 一些平台需要的数据
  options?: LXProjectOptions;
}

export interface IProjectBuilder {
  generateProject: (schema: IProjectSchema | string) => Promise<ResultDir>;
}

/** 项目级别的前置处理器 */
export type ProjectPreProcessor = (
  schema: IProjectSchema,
) => Promise<IProjectSchema> | IProjectSchema;

export interface ProjectPostProcessorOptions {
  parseResult?: IParseResult;
  template?: IProjectTemplate;
}

/** 项目级别的后置处理器 */
export type ProjectPostProcessor = (
  result: ResultDir,
  schema: IProjectSchema,
  originalSchema: IProjectSchema | string,
  options: ProjectPostProcessorOptions,
) => Promise<ResultDir> | ResultDir;

/** 模块级别的后置处理器的工厂方法 */
export type PostProcessorFactory<T> = (config?: T) => PostProcessor;

/** 模块级别的后置处理器 */
export type PostProcessor = (
  content: string,
  fileType: string,
  name?: string,
) => string;

// TODO: temp interface, need modify
export interface IPluginOptions {
  fileDirDepth: number;
}

export type BaseGenerator<I, T, C> = (
  input: I,
  scope: IScope,
  config?: C,
  next?: BaseGenerator<I, T, C>,
) => T;
type CompositeTypeGenerator<I, T> =
  | BaseGenerator<I, T, CompositeValueGeneratorOptions>
  | Array<BaseGenerator<I, T, CompositeValueGeneratorOptions>>;

export type NodeGenerator<T> = (nodeItem: IProjectSchema, scope: IScope) => T;

export interface HandlerSet<T> {
  string?: CompositeTypeGenerator<string, T>;
  boolean?: CompositeTypeGenerator<boolean, T>;
  number?: CompositeTypeGenerator<number, T>;
}

export interface CompositeValueGeneratorOptions {
  handlers?: HandlerSet<string>;
  nodeGenerator?: NodeGenerator<string>;
  tolerateEvalErrors?: boolean;
  // 指令需要操作 import
  ir?: IContainerInfo;
}

/**
 * 作用域定义，维护作用域内定义，支持作用域链上溯
 */
export interface IScope {
  // 父级作用域（如果是根作用域则为 null）
  readonly parent: IScope | null;

  // 作用域内定义的变量/函数等 bindings
  readonly bindings?: IScopeBindings;

  // TODO: 需要有上下文信息吗？ 描述什么内容
  createSubScope: (ownIndentifiers: string[]) => IScope;

  parentType?: string;
}

export interface CMDGeneratorPrames {
  value: any | any[];
  params?: any;
  platform?: string;
  scope?: IScope;
  config?: CompositeValueGeneratorOptions;
}
