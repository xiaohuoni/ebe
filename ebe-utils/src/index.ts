import { generateCode, init as ebeInit, publishers } from 'ebe';
import { type ResultDir } from 'ebe/types/core';
import { lingxiDslRules } from './constants';
import { IRulePrams, IRulesType } from './types';

export * from './types';

const isArray = (arr: any[]) => {
  if (Array.isArray) {
    return Array.isArray(arr);
  } else {
    return Object.prototype.toString.call(arr) === '[object Array]';
  }
};

export const clearProps = (data: any, diff: any) => {
  const clearedProps = { ...data };
  Object.keys(diff).forEach((key) => {
    if (typeof diff[key] === 'object' && !isArray(diff[key])) {
      clearedProps[key] = clearProps(clearedProps[key], diff[key]);
    } else if (isArray(diff[key]) && isArray(clearedProps[key])) {
      delete clearedProps[key];
    } else if (typeof diff[key] === 'function') {
      if (diff[key](clearedProps)) {
        delete clearedProps[key];
      }
    } else if (diff[key] === clearedProps[key]) {
      delete clearedProps[key];
    }
  });
  return clearedProps;
};
export const getNewDataByRules = (
  value: any,
  rule: (props: IRulePrams) => any,
  path: string[],
) => {
  const newValue = { ...value };
  let newData = newValue;
  const rootObj = rule?.({
    paths: path,
    value: newValue,
    key: 'root',
    diffraction: (diffObj) => {
      return clearProps(newValue, diffObj);
    },
  });
  if (rootObj && rootObj !== true) {
    newData = rootObj;
  } else {
    // 只支持两层
    Object.keys(newValue).forEach((newKey) => {
      const res = rule?.({
        paths: [...path, newKey],
        value: newValue[newKey],
        key: newKey,
      });
      if (res === true) {
        delete newValue[newKey];
      } else if (res) {
        newValue[newKey] = res;
      }
    });
    newData = newValue;
  }
  return newData;
};
export const removeObjectByRules = (
  data: any,
  rules: IRulesType,
  paths: string[] = [],
): any => {
  const newData: any = {};

  for (const key in data) {
    if (data.hasOwnProperty(key)) {
      const value = data[key];
      const rule = rules[key];
      const path = [...paths, key];
      if (
        rule &&
        rule.rule &&
        rule.rule({ paths: [key], value, key }) === true
      ) {
        continue;
      }
      const loopRule = rule && rule.loopRule;
      const loop = rule && rule.loop;
      if (loopRule && loop) {
        newData[key] = value.map((item: any) => {
          const loopKey = loopRule(item);
          const loopRuleObj = loop[loopKey];
          if (loopKey === 'item') {
            return removeObjectByRules(item, loopRuleObj as any, path);
          }
          if (loopRuleObj && loopRuleObj?.rule) {
            return getNewDataByRules(item, loopRuleObj?.rule, path);
          }
          return item;
        });
      } else if (typeof value === 'object' && value !== null && rule?.rule) {
        newData[key] = getNewDataByRules(value, rule?.rule, path);
      } else {
        newData[key] = value;
      }
    }
  }

  return newData;
};
// 删除 defaultVersionId 字段 ，
// components 中的 compName 值为 View 时
// 当 props 中的 conten 字段等于 $$ 时 删除 conten
// const dsRules: IRulesType = {
//   defaultVersionId: {
//     rule: () => true,
//   },
//   components: {
//     loopRule: (item) => item.compName,
//     loop: {
//       View: {
//         rule: ({ diffraction }) => {
//           return diffraction?.({
//             props: {
//               conten: '$$',
//             },
//           });
//         },
//       },
//     },
//   },
// };

export const treeForEach = (
  tree: any,
  callback: (item: any, root: boolean) => void,
  options = { children: 'components' },
) => {
  callback(tree, true);

  const loop = (children: any[]) => {
    children?.forEach((item) => {
      callback(item, false);
      if (Array.isArray(item?.[options.children])) {
        loop(item?.[options.children]);
      }
    });
  };

  const list = tree[options.children];
  loop(list);
};

export const getPageDsls = (resultObjects: any[]) => {
  return resultObjects
    .filter(Boolean)
    .map((i) => {
      try {
        const pageData = JSON.parse(i.resultObject.attrMappingJson);
        pageData.pageId = i.resultObject?.pageId;
        return pageData;
      } catch (error) {
        console.error(error, '该页面信息出错', i);
        return null;
      }
    })
    .filter(Boolean);
};
export function findAllItem<T = any>(
  target: T[],
  callback: (item: T, index: number, list: T[]) => boolean,
  itemHash: any,
) {
  const list = target;
  // Makes sures is always has an positive integer as length.
  // eslint-disable-next-line
  const length = list.length >>> 0;
  // eslint-disable-next-line
  const thisArg = arguments[1];
  for (let i = 0; i < length; ) {
    const element = list[i] as any;
    if (callback.call(thisArg, element, i, list)) {
      itemHash[element?.props?.busiCompId ?? ''] = 1;
    }
    i += 1;
    if (element?.components) {
      findAllItem(element?.components, callback, itemHash);
    }
  }
}
export function findItem<T = any>(
  target: T[],
  callback: (item: T, index: number, list: T[]) => boolean,
) {
  const list = target;
  // Makes sures is always has an positive integer as length.
  // eslint-disable-next-line
  const length = list.length >>> 0;
  // eslint-disable-next-line
  const thisArg = arguments[1];
  for (let i = 0; i < length; ) {
    const element = list[i];
    if (callback.call(thisArg, element, i, list)) {
      return element;
    }
    i += 1;
  }
  return null;
}
export const cleanTree = (tree: any, fields: any): any => {
  let fieldSet = new Set(fields); // 使用set结构可以提高查询速度

  if (Array.isArray(tree)) {
    return tree.map((item) => cleanTree(item, fields));
  } else if (typeof tree === 'object' && tree !== null) {
    return Object.entries(tree).reduce((newTree, [key, value]) => {
      if (!fieldSet.has(key)) {
        // @ts-ignore
        newTree[key] = cleanTree(value, fields);
      }
      return newTree;
    }, {});
  } else {
    return tree;
  }
};

export const clearLXPagesDSL = (pages: any[]) => {
  return pages.map((page) => {
    return removeObjectByRules(page, lingxiDslRules);
  });
};

interface CodeServices {
  qryAttrSpecPage: (data: any) => Promise<any>;
  qryPageInstListByAppId: (data: any) => Promise<any>;
  queryFrontendHookList: (data: any) => Promise<any>;
  getThemeCss: (data: any) => Promise<any>;
  qryPageCompAssetList: (data: any) => Promise<any>;
  queryFrontendDatasourcePage: (data: any) => Promise<any>;
  findPageInstByVersionId: (data: any) => Promise<any>;
  findBusiCompById: (data: any) => Promise<any>;
}

interface CodeOptions {
  appId: string;
  baseUrl: string;
  platform: string;
  services: CodeServices;
  onProgress: (data: { log: string; progress: number }) => void;
}
export const init = async () => {
  await ebeInit();
};
export const codeCreate = async ({
  appId,
  services,
  platform,
  baseUrl,
  onProgress,
}: CodeOptions) => {
  try {
    // 根据 appId 获取当前应用的全部页面
    const attrSpecPage = await services.qryAttrSpecPage({
      appId,
      pageNum: 1,
      pageSize: 999999,
    });

    // 根据 appId 获取当前应用的全部页面
    const { resultObject = [] } = await services.qryPageInstListByAppId({
      appId,
      terminalType: platform,
    });
    const { resultObject: frontendHookList = [] } =
      await services.queryFrontendHookList({
        appId,
      });

    // 根据 appId 获取当前应用的全部页面
    const themeCss = await services.getThemeCss({
      appId,
      terminalType: platform,
    });

    // 根据 appId 获取当前应用的使用的自定义组件
    const compAssetList = await services.qryPageCompAssetList({
      appId,
    });

    // 根据appId 获取全局数据源
    let globalDataInfo = await services.queryFrontendDatasourcePage({
      appId,
      pageSize: 10000,
    });

    const pageIdMapping: any = {};
    const appPageList = resultObject?.map((i: any) => {
      pageIdMapping[i.pagePath] = i.pageId;
      return i;
    });
    let lastPageId: any = '';
    // 根据 pageId 获得 dsl
    let data = [];

    data = await Promise.all(
      appPageList.map((i: any) => {
        lastPageId = i.pageId;
        return services.findPageInstByVersionId({
          appId,
          pageId: i.pageId,
        });
      }),
    );

    const pages = getPageDsls(data);
    // busiCompId 过滤重复
    const itemHash: any = {};
    // 找到所有页面使用到的 业务组件
    findAllItem(pages, (item) => item.compName === 'BOFramer', itemHash);

    const itemLists = Object.keys(itemHash);
    // 请求所有业务组件的 dsl
    const busiData = await Promise.all(
      itemLists.map((i) =>
        services.findBusiCompById({
          busiCompId: i,
          appId,
          pageId: lastPageId,
        }),
      ),
    );
    const busiCompMapping: any = {};
    const busiPages = busiData.map((i, index) => {
      const busiData = JSON.parse(
        i?.resultObject?.busiCompVersion?.sourceCodeJson,
      );
      busiData.busiCompId = i?.resultObject?.busiCompId;
      busiCompMapping[itemLists[index]] = busiData.id;
      return busiData;
    });
    // 合并页面，生成器那边支持页面类型和业务组件类型
    const pageDSLS = [...pages, ...busiPages];
    const options = {
      platform,
      appId,
      pageIdMapping,
      busiCompMapping,
      compAssetList: compAssetList?.resultObject || [],
      baseUrl,
      appConfig: {
        frontendHookList,
      },
      attrSpecPage: (attrSpecPage?.resultObject?.list || []).map(
        (i: any) => i.attrNbr,
      ),
      themeCss,
      models: globalDataInfo || {},
    };
    let cleanedTree = cleanTree(pageDSLS, ['path']); // 清理字段'b'和字段'e'
    cleanedTree = clearLXPagesDSL(cleanedTree);
    const result = await generateCode({
      solution: 'alita', // 出码方案
      options,
      schema: cleanedTree, // 编排搭建出来的 schema
      onProgress: (log: string) => {
        if (!log) {
          // 错误数据，不给返回
          return;
        }
        if (log.includes('出码生成完成')) {
          onProgress({
            log,
            progress: 100,
          });
          return;
        }
        const p = log.match(/(?<=\(整体进度: ).*?(?=\))/)?.[0] || '0';
        const pro = p.split('/') as string[];
        if (pro && pro[0] && pro[1]) {
          onProgress({
            log,
            progress: (parseInt(pro[0]) / parseInt(pro[1])) * 100,
          });
        }
      },
    });
    // console.log(result);
    // 出码结果(默认是递归结构描述的，可以传 flattenResult: true 以生成扁平结构的结果)
    publishers.zip().publish({
      project: result as ResultDir, // 上一步生成的 project
      projectSlug: appId, // 项目标识
    });
    return {
      success: true,
    };
  } catch (error) {
    return {
      msg: error,
      success: false,
    };
  }
};
