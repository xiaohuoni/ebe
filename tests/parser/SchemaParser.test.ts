import { IProjectSchema } from '../../src/core';
import { SchemaParser } from '../../src/parser/SchemaParser';
import page from '../page.json';

describe('SchemaParser', () => {
  let schemaParser: SchemaParser;

  beforeEach(() => {
    schemaParser = new SchemaParser();
  });

  describe('validate', () => {
    test('should return true', () => {
      const result = schemaParser.validate();
      expect(result).toBe(true);
    });
  });

  describe('parse', () => {
    test('should parse the schema and return the parse result', () => {
      const schemaSrc: IProjectSchema = [page] as any;
      const result = schemaParser.parse(schemaSrc);
      expect(result).toMatchSnapshot();
    });
  });
  describe('decodeSchema', () => {
    test('string schema', () => {
      const schemaSrc = '[{ "type":1}]';
      const result = schemaParser.decodeSchema(schemaSrc);
      expect(result[0].type).toBe(1);
    });
  });
  describe('decodeSchema error', () => {
    test('string schema', () => {
      const schemaSrc = '[{ type:1}]';
      expect(() => {
        schemaParser.decodeSchema(schemaSrc);
      }).toThrow();
    });
  });
});
