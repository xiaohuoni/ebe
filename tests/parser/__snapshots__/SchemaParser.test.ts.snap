// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaParser parse should parse the schema and return the parse result 1`] = `
Object {
  "app": Object {
    "keepalive": Array [],
  },
  "containers": Array [
    Object {
      "analyzeResult": Object {
        "isUsingRef": false,
      },
      "appId": "1106394452819316736",
      "components": Array [
        Object {
          "compLib": "custom",
          "compName": "View",
          "compType": 0,
          "components": Array [
            Object {
              "compLib": "comm",
              "compName": "Button",
              "compType": 1,
              "components": Array [],
              "events": Object {
                "onClick": Object {
                  "id": "Button_Button_096226_onClick",
                  "value": Array [
                    Object {
                      "callback1": Array [
                        Object {
                          "dataId": 171576713163522750,
                          "line_number": 2,
                          "options": Object {
                            "compId": "console",
                            "compName": "system",
                            "id": "179761",
                            "pageJsonId": "542158",
                            "value": Array [
                              "123",
                            ],
                          },
                          "path": undefined,
                          "shielding": undefined,
                          "type": "console",
                        },
                      ],
                      "condition": Array [
                        Object {
                          "condId": "58469",
                          "conditionType": "isVisible",
                          "objId": "Input_347757",
                          "objType": "Input",
                          "options": Object {},
                        },
                      ],
                      "dataId": 171574149702936580,
                      "elseIfs": Array [
                        Object {
                          "callback": Array [
                            Object {
                              "dataId": 171576765418323900,
                              "line_number": 3,
                              "options": Object {
                                "compId": "console",
                                "compName": "system",
                                "id": "442761",
                                "pageJsonId": "542158",
                                "value": Array [
                                  "2222",
                                ],
                              },
                              "path": undefined,
                              "shielding": undefined,
                              "type": "console",
                            },
                          ],
                          "children": Array [
                            Object {
                              "actionObjId": "console",
                              "actionObjName": "system",
                              "children": Array [],
                              "dataId": 171576765418323900,
                              "dataName": "action",
                              "elseIfs": Array [],
                              "line_number": 3,
                              "options": Object {
                                "compId": "console",
                                "compName": "system",
                                "id": "442761",
                                "pageJsonId": "542158",
                                "value": Array [
                                  "2222",
                                ],
                              },
                              "todoOptions": Array [
                                "valueArray",
                              ],
                              "value": "console",
                            },
                          ],
                          "condition": Array [
                            Object {
                              "condId": "250352",
                              "conditionType": "notVisible",
                              "objId": "Input_1814256",
                              "objType": "Input",
                              "options": Object {},
                            },
                          ],
                          "dataId": 171576763139135300,
                          "dataName": "elseIf",
                          "elseIfs": Array [],
                          "value": "elseIf",
                        },
                        Object {
                          "callback": Array [
                            Object {
                              "dataId": 171577180618680400,
                              "line_number": 4,
                              "options": Object {
                                "compId": "console",
                                "compName": "system",
                                "id": "880927",
                                "pageJsonId": "542158",
                                "value": Array [
                                  "333333",
                                ],
                              },
                              "path": undefined,
                              "shielding": undefined,
                              "type": "console",
                            },
                          ],
                          "children": Array [
                            Object {
                              "actionObjId": "console",
                              "actionObjName": "system",
                              "children": Array [],
                              "dataId": 171577180618680400,
                              "dataName": "action",
                              "elseIfs": Array [],
                              "line_number": 4,
                              "options": Object {
                                "compId": "console",
                                "compName": "system",
                                "id": "880927",
                                "pageJsonId": "542158",
                                "value": Array [
                                  "333333",
                                ],
                              },
                              "todoOptions": Array [
                                "valueArray",
                              ],
                              "value": "console",
                            },
                          ],
                          "condition": Array [],
                          "dataId": 171577179754159780,
                          "dataName": "elseIf",
                          "elseIfs": Array [],
                          "value": "elseIf",
                        },
                      ],
                      "line_number": 1,
                      "path": undefined,
                      "shielding": undefined,
                      "type": "ifelse",
                    },
                  ],
                },
              },
              "id": "Button_096226",
              "isBusiObjContainer": false,
              "isContainer": false,
              "isInlineBlock": true,
              "isLabelDropBoxChild": false,
              "label": "按钮",
              "originClassName": "",
              "platform": "pc",
              "props": Object {
                "$$componentItem": "##{{id: 'Button_096226',uid: \\"Button_096226\\"}}##",
                "block": false,
                "btnText": "按钮",
                "className": "Button_Button_096226",
                "data-compid": "Button_096226",
                "ghost": false,
                "hasIcon": false,
                "iconPosition": "left",
                "loading": false,
                "name": "按钮",
                "shape": "default",
                "size": "default",
                "type": "default",
                "visible": true,
              },
              "setEvents": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "actionObjId": "console",
                          "actionObjName": "system",
                          "children": Array [],
                          "dataId": 171576713163522750,
                          "dataName": "action",
                          "elseIfs": Array [],
                          "line_number": 2,
                          "options": Object {
                            "compId": "console",
                            "compName": "system",
                            "id": "179761",
                            "pageJsonId": "542158",
                            "value": Array [
                              "123",
                            ],
                          },
                          "todoOptions": Array [
                            "valueArray",
                          ],
                          "value": "console",
                        },
                      ],
                      "condition": Array [
                        Object {
                          "condId": "58469",
                          "conditionType": "isVisible",
                          "objId": "Input_347757",
                          "objType": "Input",
                          "options": Object {},
                        },
                      ],
                      "dataId": 171574149702936580,
                      "dataName": "condition",
                      "elseIfs": Array [
                        Object {
                          "callback": Array [
                            Object {
                              "dataId": 171576765418323900,
                              "line_number": 3,
                              "options": Object {
                                "compId": "console",
                                "compName": "system",
                                "id": "442761",
                                "pageJsonId": "542158",
                                "value": Array [
                                  "2222",
                                ],
                              },
                              "path": undefined,
                              "shielding": undefined,
                              "type": "console",
                            },
                          ],
                          "children": Array [
                            Object {
                              "actionObjId": "console",
                              "actionObjName": "system",
                              "children": Array [],
                              "dataId": 171576765418323900,
                              "dataName": "action",
                              "elseIfs": Array [],
                              "line_number": 3,
                              "options": Object {
                                "compId": "console",
                                "compName": "system",
                                "id": "442761",
                                "pageJsonId": "542158",
                                "value": Array [
                                  "2222",
                                ],
                              },
                              "todoOptions": Array [
                                "valueArray",
                              ],
                              "value": "console",
                            },
                          ],
                          "condition": Array [
                            Object {
                              "condId": "250352",
                              "conditionType": "notVisible",
                              "objId": "Input_1814256",
                              "objType": "Input",
                              "options": Object {},
                            },
                          ],
                          "dataId": 171576763139135300,
                          "dataName": "elseIf",
                          "elseIfs": Array [],
                          "value": "elseIf",
                        },
                        Object {
                          "callback": Array [
                            Object {
                              "dataId": 171577180618680400,
                              "line_number": 4,
                              "options": Object {
                                "compId": "console",
                                "compName": "system",
                                "id": "880927",
                                "pageJsonId": "542158",
                                "value": Array [
                                  "333333",
                                ],
                              },
                              "path": undefined,
                              "shielding": undefined,
                              "type": "console",
                            },
                          ],
                          "children": Array [
                            Object {
                              "actionObjId": "console",
                              "actionObjName": "system",
                              "children": Array [],
                              "dataId": 171577180618680400,
                              "dataName": "action",
                              "elseIfs": Array [],
                              "line_number": 4,
                              "options": Object {
                                "compId": "console",
                                "compName": "system",
                                "id": "880927",
                                "pageJsonId": "542158",
                                "value": Array [
                                  "333333",
                                ],
                              },
                              "todoOptions": Array [
                                "valueArray",
                              ],
                              "value": "console",
                            },
                          ],
                          "condition": Array [],
                          "dataId": 171577179754159780,
                          "dataName": "elseIf",
                          "elseIfs": Array [],
                          "value": "elseIf",
                        },
                      ],
                      "line_number": 1,
                      "value": "condition",
                    },
                  ],
                  "dataId": 171573860914746140,
                  "dataName": "event",
                  "params": Array [
                    Object {
                      "name": "e",
                      "title": "事件对象",
                      "value": "$e$",
                    },
                  ],
                  "value": "onClick",
                },
              ],
              "style": Object {
                "textAlign": "center",
              },
              "type": "Button",
              Symbol(LoopMarkSymbol): Object {
                "loopType": "",
                "parentLoopId": "",
              },
            },
            Object {
              "compLib": "comm",
              "compName": "Table",
              "compType": 1,
              "components": Array [],
              "events": Object {},
              "id": "Table_096226",
              "label": "表格",
              "originClassName": "",
              "props": Object {
                "data-compid": "Table_096226",
                "name": "表格",
              },
              "type": "Table",
              Symbol(LoopMarkSymbol): Object {
                "loopType": "",
                "parentLoopId": "",
              },
            },
          ],
          "events": Object {},
          "id": "View_542158_1",
          "isBusiObjContainer": false,
          "isContainer": true,
          "isLabelDropBoxChild": false,
          "label": "普通容器",
          "originClassName": "",
          "platform": "pc",
          "props": Object {
            "$$componentItem": "##{{id: 'View_542158_1',uid: \\"View_542158_1\\"}}##",
            "backgroundType": Object {
              "color": "rgba(255, 255, 255, 1)",
              "type": "cleanColor",
            },
            "className": "View_View_542158_1",
            "data-compid": "View_542158_1",
            "name": "页面",
            "visible": true,
          },
          "setEvents": Array [],
          "style": Object {
            "display": "flex",
            "flexDirection": "column",
            "height": "auto",
            "minHeight": "100%",
            "overflowY": "auto",
            "padding": "20px 20px 20px 20px",
            "width": "auto",
          },
          "type": "View",
          Symbol(LoopMarkSymbol): Object {
            "loopType": "",
            "parentLoopId": "",
          },
        },
      ],
      "containerType": "Page",
      "createdEditorVersion": "3.3.1.05092000",
      "currentPageVersionId": "1106394609220718593",
      "defaultVersionId": "1106394609220718593",
      "deps": Array [
        Object {
          "dependencyType": "External",
          "destructuring": true,
          "exportName": "View",
          "package": "@/components/factory",
          "type": "View",
          "version": "*",
        },
        Object {
          "dependencyType": "External",
          "destructuring": true,
          "exportName": "Button",
          "package": "@/components/factory",
          "type": "Button",
          "version": "*",
        },
        Object {
          "dependencyType": "External",
          "destructuring": true,
          "exportName": "Table",
          "package": "@/components/factory",
          "type": "Table",
          "version": "*",
        },
      ],
      "editorVersion": "3.3.1.05141705",
      "globalDataSource": Object {},
      "id": "542158",
      "imgInfoList": Array [],
      "moduleName": undefined,
      "pageId": "1106394609220718592",
      "platform": "pc",
      "platformType": "LowCode",
      "previewValidVersionId": "1106394609220718593",
      "terminalType": "PC",
      "type": "Page",
    },
  ],
  "dataSources": Array [
    Object {
      "abc": 1,
      "analyzeResult": Object {
        "isUsingRef": false,
      },
      "containerType": "Page",
      "dataSource": "[]",
      "moduleName": undefined,
      "type": "Page",
    },
  ],
  "globalDatas": Array [],
  "models": Object {},
  "pageview": Object {
    "models": Array [],
    "pageStaticData": Object {},
    "routes": Array [
      Object {
        "fileName": undefined,
        "pageId": undefined,
        "path": undefined,
        "type": undefined,
      },
    ],
  },
  "project": Object {
    "containersDeps": Array [],
    "packages": Array [
      Object {
        "package": "@/components/factory",
        "version": "*",
      },
    ],
  },
  "staticFiles": undefined,
}
`;
